heat/core/linalg/bcg.py

```diff
@@ -100,59 +107,338 @@ def bi_diagonalize(A, overwrite_arr=True):
 
         B : ht.DNDarray
 
+
+    bl -> The lower bandwidth of the matrix.
+    bu -> The upper bandwidth of the matrix.
+    b = bl + bu => Is the number of off diagonals present in the matrix.
+
+    The reduction proceeds in k = min(m,n) steps each of which produces one row of the resulting bidiagonal matrix B.
+    The band is partitioned into block columns Dj,Ej and transformations are applied in each sweep.
+
+    After the kth sweep, the leading k X (k + 1) block (Bk) of A contains the first k rows of the resulting bidiagonal matrix B.
+    There is a slight difference in the way band is partitioned into block columns Dj,Ej for the cases where m>=n & n>m.
+
+
+
+    To change the serial algorithm into an Efficient parallel program, We should distribute the data and the computational work to multiple processors.
+
+    The band form is further reduced to the final condensed form using the bulge chasing technique.
+    This procedure annihilates the extra off-diagonal elements by chasing the created
+    fill-in elements down to the bottom right side of the matrix using successive
+    orthogonal transformations at each sweep.
+
+
+
+    Note: We will do the bulge chasing using Halos.
+
+    Second Stage of reduction: Band matrix to a real bidiagonal matrix. This uses a bulge chasing algorithm.
+
+
     """
     if overwrite_arr:
         arr = A
         # The input matrix is overwritten with the result, i.e it will be changed to a bidiagonal matrix.
     else:
         arr = A.copy()
-        # The input matrix is not overwritten, i.e a new matrix "arr" which is a copy of input matrix, will be changed to a bidiagonal matrix.
+        # The input matrix is not overwritten, i.e a new matrix "arr" which is a copy of the input matrix, will be changed to a bidiagonal matrix.
 
     m, n = arr.shape
+    k = min(m, n)
+    # k is the minimum of m and n
+
+    # Find the width of the diagonal of the input matrix.
+
+    b = 0
+    bl = 0
+    bu = 0
+    row_0 = arr[0, 0:]
+    col_0 = arr[0:, 0]
+
+    for i in range(1, len(col_0)):
+        if col_0[i] != 0:
+            bl += 1
+
+    for i in range(1, len(row_0)):
+        if row_0[i] != 0:
+            bu += 1
+
+    b = bl + bu
+
+    # print("b: ", b)
+
+    U1, vt1 = ht.eye(m, dtype=ht.float64), ht.eye(n, dtype=ht.float64)
```

Comment:
since we dont know what that the input will be a float64 matrix, you should use the dtype of the input DNDarray (`arr.dtype`)

Reply:
Changed it to arr.dtype